generators/json_generators.py

```diff
@@ -0,0 +1,235 @@
+
+        Args:
+            key: Key value in the JSON file.
+            fn_name: Function name ; String
+            args : Function arguments as a dictionary.
+
+        Example usage:
+
+            args = {
+                "parameter_1" : "value_1",
+                "parameter_2" : [28, 28, 1]
+            }
+            add_fn("key", "function", args)
+        '''
+        # Raise exceptions if args incorrectly formatted
+        generator_utils._is_valid_arg_dict(args)
+
+        _fn = {
+            "name": fn_name,
+            "args": args
+        }
+        self.add_entry(key, _fn)
+
+    def write(self):
+        ''' Writes the root dictionary to the JSON file and closes the file.
+        '''
+        json.dump(self.root, self.out, indent=4)
+        self._close()
+        print("Saved to {}".format(self.out_file))
+
+
+class PipelineJsonGenerator(JsonGenerator):
+    ''' Generates a dataset pipeline configuration file based on user selections..
+    '''
+
+    def __init__(self, out_file):
+
+        super(PipelineJsonGenerator, self).__init__(out_file)
+
+        # Add the root dict
+        self.add_entry("pipeline", {})
+
+        # Add the dataset and operations
+        self._indent("pipeline")
+        self.add_entry("dataset", {})
+        self.add_entry("operations", [])
+
+    def add_dataset(self, label):
+        ''' Add a dataset source.
+
+        All available datasets can be found in the Tensorflow Datasets catalog.
+            (https://www.tensorflow.org/datasets/catalog/overview)
+
+        Args:
+            label : dataset identifier. Equivalent to the Tensorflow dataset name.
+        '''
+
+        self.add_entry("dataset", {"label": label})
+
+    def add_operation(self, op_name, args=None):
+        ''' Add a preprocessing operation.
+
+        A list of all allowable operations can be found as methods for the tf.data.Dataset class
+            (https://www.tensorflow.org/api_docs/python/tf/data/Dataset)
+
+        Args:
+            op_name : Name of the operation. Equivalent to a tf.Dataset method name.
+            args : Dictionary of arguments (param : value) passed to the method. The value doesn't
+                always correspond to the actual argument so that functionality can be abstracted
+                from specific machine learning libraries. Check variable_map.json for all values
+                and their representations.
+
+            Example usage:
+
+            args = {
+                    "map_func" : "normalize_img",
+                    "num_parallel_calls" : "autotune"
+                    }
+
+            pipeline.add_operation("map", args)
+        '''
+        self.add_fn("operations", op_name, args)
+
+    def get_dataset_list(self):
+        '''Return a list of all possible dataset sources.
+        '''
+        pass
+
+    def get_operations_list(self):
+        ''' Return a list of all possible preprocessing operations.
+        '''
+        pass
```

Comment:
What are you imagining these will be used for?

Reply:
Down the road, the user will need to know what options are available for parameters to a block function. It would be nice to have them retrievable instead of them having to scour the documentation

Reply:
Oh I see, so the JSON generator is the API that will directly be invoked by the front-end UI?

Reply:
yes